internal/common/retry/interceptor.go

```diff
@@ -0,0 +1,156 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package retry
+
+import (
+	"context"
+	"math"
+	"time"
+
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/backoffutils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	// UnlimitedMaximumAttempts when maximum attempts is set to this special value, then the number of attempts is unlimited.
+	UnlimitedMaximumAttempts = 0
+	// UnlimitedInterval when maximum interval is set to this special value, then there is no upper bound on the retry delay.
+	// Should not be used together with unlimited attempts as resulting retry interval can grow to unreasonable values.
+	UnlimitedInterval = 0
+	// DefaultBackoffCoefficient is default backOffCoefficient for retryPolicy
+	DefaultBackoffCoefficient = 2.0
+	// DefaultMaximumInterval is default maximum amount of time for an individual retry.
+	DefaultMaximumInterval = 10 * time.Second
+	// DefaultExpirationInterval is default expiration time for all retry attempts.
+	DefaultExpirationInterval = time.Minute
+	// DefaultMaximumAttempts is default maximum number of attempts.
+	DefaultMaximumAttempts = UnlimitedMaximumAttempts
+	// DefaultJitter is a default jitter applied on the backoff interval for delay randomization.
+	DefaultJitter = 0.2
+)
+
+type (
+	// GrpcRetryConfig defines required configuration for exponential backoff function that is supplied to gRPC retrier.
+	GrpcRetryConfig struct {
+		initialInterval    time.Duration
+		backoffCoefficient float64
+		maximumInterval    time.Duration
+		expirationInterval time.Duration
+		jitter             float64
+		maximumAttempts    int
+	}
+
+	contextKey string
+)
+
+func (ck contextKey) String() string {
+	return string(ck)
+}
```

Comment:
```suggestion
```

Reply:
I would prefer to keep string value for nicer debugging.